doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,286 @@
+2. Harry Potter World
+3. Pottery barn
+4. Harry Potter
+
+These are not good results.
+
+Our approach now (prior to implementing Learning To Rank) would be to
+manually rank these results, giving them a score of between 0-3 with
+0 being irrelevant and 3 being completely relevant:
+
+```
+query         rating    document
+harry potter  0         HMRC Sign in
+harry potter  2         Harry Potter World
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+```
+
+These judgements are an expression of the ideal ordering.
+
+We can then use nDCG to tell us how good the results are.
+We might get a score of 0.3 for this result set (from 0 to 1).
+
+We could then investigate why the results are bad. Perhaps there is a
+best bet. Or popularity might be pushing HMRC docs. We can fix these
+things, then re-run nDCG, and we might get 0.7 for this query.
+
+In contrast, with Learning To Rank, search result ranks would be
+determined by a model without manually changing the boost of a field.
+
+With Learning To Rank we would go through the same process of
+rating the search results.
+
+```
+harry potter  0         Pottery barn
+harry potter  3         Harry Potter
+...
+```
+
+However, after doing this, we would train a model with this data
+to rank the results better. The model optimises for nDCG.
+
+We provide the model with features, such as the fields we boost on.
+
+```
+query         rating    document      title_score view_count  recency  pagerank ...
+harry potter  0         Pottery barn  0.2         9876        0.12     9876     ...
+harry potter  3         Harry Potter  0.2         1234        1234     1234     ...
+...
+```
+
+For example, take the SVM dataset below for the query 'harry potter' (qid:1)
+
+```
+0 qid:1 popularity:0.3 title_score:0.2 description_score:0.3 recency:0.9 pagerank:5 ...
+3 qid:1 popularity:0.1 title_score:0.6 description_score:0.7 recency:0.5 pagerank:2 ...
+```
+
+These graded query:document pairs and their features can be used
+to train a model.
+
+The model might learn that title and description scores are slightly
+more useful that popularity and recency, and how much, in certain
+circumstances.
+
+Rather than manually tuning boosts, a model learns from relevancy
+judgements the role that features should play when providing relevant results.
+
+This is a high level overview of what Learning To Rank is for our use case.
+There has been a lot of research in this area, and within the fields of
+information retrieval & machine learning which Learning To Rank intersects.
+There is a lot more out there to learn!
+
+For more details about how we plan to implement Learning To Rank, see
+the implementation section below.
+
+## Motivation
+
+We have decided to try Learning To Rank because it has the potential
+to greatly improve search relevance in a way that is:
+
+- automatable
+- scalable
+- more maintainable
+- more successful
+
+**Automatable** because the model won't require human involvement to
```

Comment:
(And figuring out how to react to them)

Reply:
That's a good point, I think you're right. I've added a task to the epic to investigate this.

Reply:
Is it fair to say that automatable is directly relevant to accuracy also? If we remove the human involvement then we hopefully make this more accurate, which is the point. This is quite similar to successful, but thought I'd add this anyway.

Reply:
> Is it fair to say that automatable is directly relevant to accuracy also? 

Yes I think so. Rather than requiring a human to make a change, we can have the model use the latest user data and automatically adjust the way it ranks results. For example, boosting factors are arbitrary, can get out of date, and may need manually changing in future, requiring human involvement. But this wouldn't happen with LTR because the model will automatically compute the optimal weighting each field should have for a given query.

There are also some things that we won't be able to do manually to improve relevancy, but that the reranker model can, one example: https://github.com/alphagov/search-api/pull/1778/files#diff-5ea6bc45fb5fedbdaa06cec834586979R155

> ensuring the judgements it's trained on are up to date and representative

I agree!

> Maybe there are particular topics/search terms where you want more control/oversight

Yes, one thing we are thinking about is how we can allow editors to control the algorithm.

My thought at the beginning of the project was that we would use both manual and autogenerated relevancy judgements to train the model. This would let us manually influence the model - @hannako is working on a prototype for this.

But a month of user data can generate more than a million relevancy judgements. So the manual judgements might get drowned out. I was thinking perhaps we could use manual judgements in a final validation stage so we could spot any queries that are performing badly.

We could also use the model-generated scores as a multiplicative factor, so that best bets continue to work, and use something like best bets as a way to curate results.

> How do you account for bot activity elsewhere on GOV.UK?

We use javascript for collecting implicit judgements, which removes some bot activity. We could definitely look into this, there's likely to be monitoring around this that I'm not aware of.

Another note is that a bot viewing and clicking a lot on a favoured result for a particular query won't necessarily result in that result being ranked higher for that query by the model. That behaviour would produce a set of relevancy judgements for that query, which would be perhaps 20 judgements in a set of millions. I think gaming the model to lead to a particular outcome would be tricky, as it would require some sophistication and a lot of activity that we might notice.

> What automated quality checks could you run on newly trained models to prevent broken models getting into production?

I'm hoping we can re-use the approach we've had for manual changes for shipping new models (offline ranking evaluation metrics followed by online AB tests).

At the moment we've been using nDCG to compare models with the previous version (right now we just compare it to our existing search). The ranking model also records NDCG@1,3,5,10 and other metrics, so we can see if it is better or worse for the data we've given it. We can also run a set of manual relevancy judgements - that we know are good quality - against the model, to see if it gets better for those judgements (I've been doing this when working on the model).

> Are particular audiences over/under represented when you are producing the judgments? You can potentially weight clicks differently based on demographic or device type.

We normalise judgements to scores between 0-3, so judgements for highly popular queries have the same weighting for very low popularity pages (e.g. a query for "whitechapel building" that gets 20 hits has the same level of impact as the judgements for "brexit").

I think this is quite a nice feature of the model, since it treats the judgements of queries equally, regardless of popularity or how they were generated.

However, if we have a niche kind of query, with not many judgements for that type, then the results for that query type could get worse.

For example, searches for people are relatively not very popular. The model (arguably) struggles a bit with search for names at the moment. It favours articles by/about people rather than the canonical person pages.

We'll need to find a way to solve this problem, so maybe manually creating relevancy judgements for these underrepresented queries would be the answer. Weighting clicks/judgements is an interesting idea. I'm not sure I know how we'd do this at the moment - there's a card in our backlog to look into how we can react to underperforming queries.